# Cyclist data analysis
----------------------------------

In [1]:
import pandas as pd

In [2]:
q1_2019 = pd.read_csv("Divvy_trips_data/Divvy_Trips_2019_Q1.csv")
q1_2020 = pd.read_csv("Divvy_trips_data/Divvy_Trips_2020_Q1.csv")
print("Data loaded successfully!")

Data loaded successfully!


#============================================
#### Wrangling & Combining Data into a Single File
#============================================

In [3]:
# Checking the columns for each df
print("Columns for Q1 2019:\n", q1_2019.columns)
print("\nColumns for Q1 2020:\n", q1_2020.columns)

Columns for Q1 2019:
 Index(['trip_id', 'start_time', 'end_time', 'bikeid', 'tripduration',
       'from_station_id', 'from_station_name', 'to_station_id',
       'to_station_name', 'usertype', 'gender', 'birthyear'],
      dtype='object')

Columns for Q1 2020:
 Index(['ride_id', 'rideable_type', 'started_at', 'ended_at',
       'start_station_name', 'start_station_id', 'end_station_name',
       'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng',
       'member_casual'],
      dtype='object')


In [4]:
# Renaming the columns in q1_2019 to match the column names in q1_2020 for consistency
q1_2019 = q1_2019.rename(columns={
    'trip_id':'ride_id',
    'bikeid':'rideable_type',
    'start_time':'started_at',
    'end_time':'ended_at',
    'from_station_name':'start_station_name',
    'from_station_id':'start_station_id',
    'to_station_name':'end_station_name',
    'to_station_id':'end_station_id',
    'usertype':'member_casual'
})

# Inspecting the data frames to make sure everything correct.
print("\nQ1 2019 Data info:")
q1_2019.info()
print("\nQ1 2020 Data info:")
q1_2020.info()



Q1 2019 Data info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365069 entries, 0 to 365068
Data columns (total 12 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             365069 non-null  int64  
 1   started_at          365069 non-null  object 
 2   ended_at            365069 non-null  object 
 3   rideable_type       365069 non-null  int64  
 4   tripduration        365069 non-null  object 
 5   start_station_id    365069 non-null  int64  
 6   start_station_name  365069 non-null  object 
 7   end_station_id      365069 non-null  int64  
 8   end_station_name    365069 non-null  object 
 9   member_casual       365069 non-null  object 
 10  gender              345358 non-null  object 
 11  birthyear           347046 non-null  float64
dtypes: float64(1), int64(4), object(7)
memory usage: 33.4+ MB

Q1 2020 Data info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426887 entries, 0 to 426886
Data col

In [5]:
# Converting the ride_id and rideable_type columns to string type for consistency
q1_2019["ride_id"] = q1_2019["ride_id"].astype(str)
q1_2019["rideable_type"] = q1_2019["rideable_type"].astype(str)

In [6]:
# Stacking the two dataframes together to create a single dataframe for analysis
all_trips = pd.concat([q1_2019, q1_2020], ignore_index=True)

In [7]:
# Dropping unnecessary columns that are not needed for the analysis
all_trips = all_trips.drop(columns=['start_lat', 'start_lng', 'end_lat', 'end_lng', 'birthyear', 'gender', 'tripduration'], errors='ignore')

#===========================================
### Cleaning and Adding Data for Analysis
#===========================================

In [8]:
# inspecting the combined dataframe to make sure everything is correct.
print("\nList of columns after combining & dropping unnecessary columns:\n", all_trips.columns)
print("\nHow many rows are in the combined dataframe:\n", len(all_trips))
print("\nDimensions of the dataframe (rows, columns):\n", all_trips.shape)
print("\nfirst 6 rows of the combined dataframe:\n", all_trips.head())
print("\nList of columns and data types:\n")
all_trips.info()

#
print("\nStatistical summary of the dataframe:\n", all_trips.describe(include='all'))


List of columns after combining & dropping unnecessary columns:
 Index(['ride_id', 'started_at', 'ended_at', 'rideable_type',
       'start_station_id', 'start_station_name', 'end_station_id',
       'end_station_name', 'member_casual'],
      dtype='object')

How many rows are in the combined dataframe:
 791956

Dimensions of the dataframe (rows, columns):
 (791956, 9)

first 6 rows of the combined dataframe:
     ride_id           started_at             ended_at rideable_type  \
0  21742443  2019-01-01 00:04:37  2019-01-01 00:11:07          2167   
1  21742444  2019-01-01 00:08:13  2019-01-01 00:15:34          4386   
2  21742445  2019-01-01 00:13:23  2019-01-01 00:27:12          1524   
3  21742446  2019-01-01 00:13:45  2019-01-01 00:43:28           252   
4  21742447  2019-01-01 00:14:52  2019-01-01 00:20:56          1170   

   start_station_id                   start_station_name  end_station_id  \
0               199               Wabash Ave & Grand Ave            84.0   
1    

In [9]:
# Checking how many observations are in each category of the member_casual column
print("\nValue counts for the member_casual column:\n", all_trips["member_casual"].value_counts())


Value counts for the member_casual column:
 member_casual
member        378407
Subscriber    341906
casual         48480
Customer       23163
Name: count, dtype: int64


In [10]:
# replacing the values in the member_casual column to make them more readable and easier to understand
all_trips["member_casual"] = all_trips["member_casual"].replace({"Subscriber": "member", "Customer": "casual"})

In [11]:
# checking to make sure the proper number of observations are in each category of the member_casual column after the replacement
print("\nValue counts for the member_casual column:\n", all_trips["member_casual"].value_counts())


Value counts for the member_casual column:
 member_casual
member    720313
casual     71643
Name: count, dtype: int64


In [12]:
# Converting the 'started_at' and 'ended_at' columns to datetime format
all_trips["started_at"] = pd.to_datetime(all_trips["started_at"])

all_trips["ended_at"] = pd.to_datetime(all_trips["ended_at"])


In [13]:
all_trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 791956 entries, 0 to 791955
Data columns (total 9 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             791956 non-null  object        
 1   started_at          791956 non-null  datetime64[ns]
 2   ended_at            791956 non-null  datetime64[ns]
 3   rideable_type       791956 non-null  object        
 4   start_station_id    791956 non-null  int64         
 5   start_station_name  791956 non-null  object        
 6   end_station_id      791955 non-null  float64       
 7   end_station_name    791955 non-null  object        
 8   member_casual       791956 non-null  object        
dtypes: datetime64[ns](2), float64(1), int64(1), object(5)
memory usage: 54.4+ MB


In [14]:
# add columns that list the date, month, day, year, and day of week of each ride
all_trips["date"] = all_trips["started_at"].dt.date
all_trips["month"] = all_trips["started_at"].dt.month
all_trips["day"] = all_trips["started_at"].dt.day
all_trips["year"] = all_trips["started_at"].dt.year
all_trips["day_of_week"] = all_trips["started_at"].dt.day_name()


In [15]:
# Calculate the length of each ride in seconds
all_trips["ride_length"] = (all_trips["ended_at"] - all_trips["started_at"]).dt.total_seconds()

# Display information about the DataFrame including the new columns
print("\nData types after adding new columns:\n")
all_trips.info()


Data types after adding new columns:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 791956 entries, 0 to 791955
Data columns (total 15 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             791956 non-null  object        
 1   started_at          791956 non-null  datetime64[ns]
 2   ended_at            791956 non-null  datetime64[ns]
 3   rideable_type       791956 non-null  object        
 4   start_station_id    791956 non-null  int64         
 5   start_station_name  791956 non-null  object        
 6   end_station_id      791955 non-null  float64       
 7   end_station_name    791955 non-null  object        
 8   member_casual       791956 non-null  object        
 9   date                791956 non-null  object        
 10  month               791956 non-null  int32         
 11  day                 791956 non-null  int32         
 12  year                791956 non-null  int32     

In [16]:
# checking for trips with ride lengths negative or zero
all_trips[all_trips["ride_length"] <= 0]

,ride_id,started_at,ended_at,rideable_type,start_station_id,start_station_name,end_station_id,end_station_name,member_casual,date,month,day,year,day_of_week,ride_length
519241,23EF1DCC9FCA40BA,2020-02-28 11:34:40,2020-02-28 11:34:40,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-02-28,2,28,2020,Friday,0.0
520401,9461DFF13D8BA8AD,2020-02-28 10:09:43,2020-02-28 10:09:42,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-02-28,2,28,2020,Friday,-1.0
527968,86163D9676BBBE62,2020-02-26 14:41:16,2020-02-26 14:41:16,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-02-26,2,26,2020,Wednesday,0.0
536467,836931C569802344,2020-02-27 09:56:47,2020-02-27 09:56:47,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-02-27,2,27,2020,Thursday,0.0
537472,07CD3CBC94106B37,2020-02-28 10:02:30,2020-02-28 10:02:30,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-02-28,2,28,2020,Friday,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
788789,09C4B33945876A83,2020-03-10 09:54:23,2020-03-10 09:54:23,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-03-10,3,10,2020,Tuesday,0.0
788945,323CAC41316F59B2,2020-03-13 08:51:45,2020-03-13 08:51:34,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-03-13,3,13,2020,Friday,-11.0
789251,476717BBABCAE518,2020-03-19 09:31:56,2020-03-19 09:31:45,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-03-19,3,19,2020,Thursday,-11.0
790218,F809003AA48EC0C1,2020-03-09 12:12:59,2020-03-09 12:12:59,docked_bike,675,HQ QR,675.0,HQ QR,casual,2020-03-09,3,9,2020,Monday,0.0


In [17]:
#* checking for trips with ride lengths greater than 24 hours (1440 minutes)
all_trips[all_trips["ride_length"] > 1440]

,ride_id,started_at,ended_at,rideable_type,start_station_id,start_station_name,end_station_id,end_station_name,member_casual,date,month,day,year,day_of_week,ride_length
3,21742446,2019-01-01 00:13:45,2019-01-01 00:43:28,252,123,California Ave & Milwaukee Ave,176.0,Clark St & Elm St,member,2019-01-01,1,1,2019,Tuesday,1783.0
8,21742451,2019-01-01 00:18:43,2019-01-01 00:47:30,6205,150,Fort Dearborn Dr & 31st St,148.0,State St & 33rd St,member,2019-01-01,1,1,2019,Tuesday,1727.0
19,21742463,2019-01-01 00:29:19,2019-01-01 01:08:12,3914,35,Streeter Dr & Grand Ave,39.0,Wabash Ave & Adams St,casual,2019-01-01,1,1,2019,Tuesday,2333.0
21,21742465,2019-01-01 00:29:28,2019-01-01 01:07:49,3355,35,Streeter Dr & Grand Ave,39.0,Wabash Ave & Adams St,casual,2019-01-01,1,1,2019,Tuesday,2301.0
30,21742475,2019-01-01 00:45:37,2019-01-01 01:17:51,5583,464,Damen Ave & Foster Ave,30.0,Ashland Ave & Augusta Blvd,member,2019-01-01,1,1,2019,Tuesday,1934.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
791942,46230CF6653E2CDE,2020-03-25 17:59:02,2020-03-25 18:29:31,docked_bike,459,Lakefront Trail & Bryn Mawr Ave,313.0,Lakeview Ave & Fullerton Pkwy,member,2020-03-25,3,25,2020,Wednesday,1829.0
791948,4361A2348584C3E4,2020-03-25 17:10:10,2020-03-25 17:47:34,docked_bike,313,Lakeview Ave & Fullerton Pkwy,459.0,Lakefront Trail & Bryn Mawr Ave,member,2020-03-25,3,25,2020,Wednesday,2244.0
791952,82B10FA3994BC66A,2020-03-07 15:25:55,2020-03-07 16:14:03,docked_bike,161,Rush St & Superior St,240.0,Sheridan Rd & Irving Park Rd,member,2020-03-07,3,7,2020,Saturday,2888.0
791953,AA0D5AAA0B59C8AA,2020-03-01 13:12:38,2020-03-01 13:38:29,docked_bike,141,Clark St & Lincoln Ave,210.0,Ashland Ave & Division St,casual,2020-03-01,3,1,2020,Sunday,1551.0


In [18]:
# Removing the rows with negative ride lengths and rides that start at "HQ QR" by filtering the dataframe and creating a new dataframe called all_trips_v2 using copy() to avoid the SettingWithCopyWarning. This will ensure that the original dataframe remains unchanged and we can work with a clean dataset for analysis.
all_trips_v2 = all_trips[(all_trips["start_station_name"] != "HQ QR") & (all_trips["ride_length"].between(0, 1440))].copy()

#============================
### DESCRIPTIVE ANALYSIS
#============================

In [19]:
# Descriptive analysis of the ride_length column to understand the distribution of ride lengths in the dataset to identify any potential outliers or anomalies in the data that may need to be addressed before further analysis.
print("\nDescriptive statistics for ride_length column:\n", all_trips_v2["ride_length"].describe())


Descriptive statistics for ride_length column:
 count    699686.000000
mean        558.674686
std         318.391598
min           1.000000
25%         310.000000
50%         486.000000
75%         748.000000
max        1440.000000
Name: ride_length, dtype: float64


The descriptive statistics for ride_length show clear anomalies: the maximum value exceeds 10 million seconds (~123 days) and the standard deviation is far larger than the mean. 

These patterns indicate the presence of extreme outliers or erroneous ride duration records. 

In [20]:
# grouping the data by member_casual and calculating the mean, median, max, and min ride lengths for each group to understand how the ride lengths differ between members and casual riders.
print("\nRide Length statistics grouped by member_casual:\n", all_trips_v2.groupby('member_casual')['ride_length'].agg(['mean','median','max','min']))


Ride Length statistics grouped by member_casual:
                      mean  median     max  min
member_casual                                 
casual         811.825058   800.0  1440.0  2.0
member         545.246027   474.0  1440.0  1.0


In [21]:
# Average ride length by member type and day of the week
print("\nAverage ride length by member type and day of the week:\n")
all_trips_v2.groupby(['member_casual','day_of_week'])['ride_length'].mean()


Average ride length by member type and day of the week:



member_casual  day_of_week
casual         Friday         778.349756
               Monday         768.478052
               Saturday       855.710136
               Sunday         868.067888
               Thursday       776.691848
               Tuesday        770.209871
               Wednesday      780.602524
member         Friday         536.319190
               Monday         542.875007
               Saturday       551.887539
               Sunday         564.175615
               Thursday       542.326339
               Tuesday        546.860427
               Wednesday      545.172500
Name: ride_length, dtype: float64

Across both groupby analyses (ride length statistics by member_casual, and average ride length by member_casual and day_of_week), the results show a clear data‑quality issue in the ride_length column for casual riders.

Member ride durations fall within realistic ranges (mean around 13 minutes, median around 8 minutes). In contrast, casual riders show extremely inflated values, including a maximum ride length of 10,632,022 seconds (~123 days) and mean values exceeding 5,300 seconds (~1.5 hours). These durations are not plausible for bike‑share usage and strongly indicate the presence of corrupted or erroneous records (e.g., trips not properly ended, timestamp errors, or multi‑day rides).

These outliers significantly distort the mean and max values for casual riders in both groupby results. Before drawing conclusions or comparing member vs casual behavior, the dataset should be cleaned by removing or correcting unrealistic ride durations.

In [22]:
# Average ride length by member type and day of the week
print("\nAverage ride length by member type and day of the week:\n")
avg_ride_length = all_trips_v2.groupby(['member_casual','day_of_week'])['ride_length'].mean()

avg_ride_length


Average ride length by member type and day of the week:



member_casual  day_of_week
casual         Friday         778.349756
               Monday         768.478052
               Saturday       855.710136
               Sunday         868.067888
               Thursday       776.691848
               Tuesday        770.209871
               Wednesday      780.602524
member         Friday         536.319190
               Monday         542.875007
               Saturday       551.887539
               Sunday         564.175615
               Thursday       542.326339
               Tuesday        546.860427
               Wednesday      545.172500
Name: ride_length, dtype: float64

In [23]:
# Ordering the days of the week to ensure that the analysis is done in a logical order from Monday to Sunday. This will help in visualizing and interpreting the data more effectively.
days_order = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
all_trips_v2["day_of_week"] = pd.Categorical(all_trips_v2["day_of_week"], categories=days_order, ordered=True)



In [24]:
# Running again the aggregation to see the average ride length by member type and day of the week after ordering the days of the week.
print("\nAverage ride length by member type and day of the week:\n", all_trips_v2.groupby(['member_casual','day_of_week'], observed=True)['ride_length'].mean())


Average ride length by member type and day of the week:
 member_casual  day_of_week
casual         Monday         768.478052
               Tuesday        770.209871
               Wednesday      780.602524
               Thursday       776.691848
               Friday         778.349756
               Saturday       855.710136
               Sunday         868.067888
member         Monday         542.875007
               Tuesday        546.860427
               Wednesday      545.172500
               Thursday       542.326339
               Friday         536.319190
               Saturday       551.887539
               Sunday         564.175615
Name: ride_length, dtype: float64


In [25]:
# Calculating summary statistics by member type and day of the week, including the number of rides and average duration of rides for each group. This will provide insights into the riding patterns of members and casual riders throughout the week.
summary_stats = all_trips_v2.groupby(['member_casual','day_of_week'], observed=True).agg(
    number_of_rides = ('ride_id', 'count'),
    average_duration = ('ride_length', 'mean')
    ).reset_index()


print("\nSummary statistics by member type and day of the week:\n", summary_stats)


Summary statistics by member type and day of the week:
    member_casual day_of_week  number_of_rides  average_duration
0         casual      Monday             3326        768.478052
1         casual     Tuesday             4336        770.209871
2         casual   Wednesday             3882        780.602524
3         casual    Thursday             4183        776.691848
4         casual      Friday             4709        778.349756
5         casual    Saturday             6679        855.710136
6         casual      Sunday             8131        868.067888
7         member      Monday           102638        542.875007
8         member     Tuesday           118612        546.860427
9         member   Wednesday           113003        545.172500
10        member    Thursday           116425        542.326339
11        member      Friday           107331        536.319190
12        member    Saturday            53423        551.887539
13        member      Sunday            53008  

In [26]:
# Saving the summary statistics to a CSV file for further analysis or reporting.
counts = all_trips_v2.groupby(['member_casual',
'day_of_week'], observed=True)['ride_length'].mean().reset_index()
counts.to_csv('avg_ride_length_v3.csv', index=False)


counts

,member_casual,day_of_week,ride_length
0,casual,Monday,768.478052
1,casual,Tuesday,770.209871
2,casual,Wednesday,780.602524
3,casual,Thursday,776.691848
4,casual,Friday,778.349756
5,casual,Saturday,855.710136
6,casual,Sunday,868.067888
7,member,Monday,542.875007
8,member,Tuesday,546.860427
9,member,Wednesday,545.172500
